File: src/main/java/de/flapdoodle/generics/selftypes/j/Sample.java
package de.flapdoodle.generics.selftypes.j;

import java.util.List;

public class Sample implements SampleType<Sample> {

  private final String name;
  private final List<? extends SampleType<?>> children;

  public Sample(String name, List<? extends SampleType<?>> children) {
    this.name = name;
    this.children = children;
  }

  @Override
  public Sample changeName(String name) {
    return new Sample(name, this.children);
  }

  @Override
  public String name() {
    return name;
  }

  @Override
  public List<? extends SampleType<?>> children() {
    return children;
  }
}

File: src/main/kotlin/de/flapdoodle/generics/selftypes/k/Sample.kt
package de.flapdoodle.generics.selftypes.k

class Sample(
    private val name: String,
    private val children: List<SampleType<*>>
) : SampleType<Sample> {

  override fun changeName(name: String): Sample {
    return Sample(name, this.children)
  }

  override fun name(): String {
    return name
  }

  override fun children(): List<SampleType<*>> {
    return children
  }
}
